[database] Set tvshow_view return maximum path

wildcat2020 opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->
Added subquery in FROM clause in place of tvshowlinkpath table to return only the maximum strPath.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->
This change is required following PR #15185 in order not to break Kodi functionality that expects just one line per tvshow (as paths can be more than one)

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->
Sanity checks performed. View is operational.

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

wildcat2020 commented:
Tested latest build and views are working ok from within MySQL.

![image](https://user-images.githubusercontent.com/23155784/50727819-06631980-1129-11e9-8558-32057771e4c9.png)

pkerling commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -399,9 +399,17 @@ void CVideoDatabase::CreateViews()
+                                     "  tvshow.userrating AS userrating,"
+                                     "  tvshow.duration AS duration,"
+                                     "  min(path.idParentPath) AS idParentPath,"
+                                     "  min(path.strPath) AS strPath,"
```
Is `min` necessary here? Shouldn't `strPath` be uniquely defined by the join?

wildcat2020 replied:
Hi, unfortunately not. Remember what we discussed in PR #15185 
table path may hold more than one rows for each tvshow, therefore we need to get whichever column is selected from table path into an aggregate value in order to simulate the arbitrary random row selected if we just used group by tvshow.idShow alone (with not a full group by clause that is). We need to make sure only one row is selected per tvshow so that we don't break current Kodi functionality.
I hope that makes sense.

phate89 replied:
there's a different problem. I completely forgotten we expose also idparentpath.. that might be a problem since now we could have a path and a idparentpath from a different path.
I looked where idparentpath is used (i'm not 100% sure though) and with the tvshow_view is only used here:
https://github.com/xbmc/xbmc/blob/087521b803628221fda45626a8cca9809663c66b/xbmc/video/VideoDatabase.cpp#L10065
So we might actually drop that field from tvshow_view and fix the filter in that query..
Or we could simply change the tvshow_view to do the grouping in a subquery and remove it from the main level like this
```
SELECT  
  tvshow.idShow AS idShow, 
  tvshow.c00, tvshow.c01, tvshow.c02, tvshow.c03, 
  tvshow.c04, tvshow.c05, tvshow.c06, tvshow.c07, 
  tvshow.c08, tvshow.c09, tvshow.c10, tvshow.c11, 
  tvshow.c12, tvshow.c13, tvshow.c14, tvshow.c15, 
  tvshow.c16, tvshow.c17, tvshow.c18, tvshow.c19, 
  tvshow.c20, tvshow.c21, tvshow.c22, tvshow.c23, 
  tvshow.userrating AS userrating, 
  tvshow.duration AS duration, 
  path.idParentPath AS idParentPath, 
  path.strPath AS strPath, 
  tvshowcounts.dateAdded AS dateAdded, 
  lastPlayed, totalCount, watchedcount, totalSeasons,  
  rating.rating AS rating,  
  rating.votes AS votes,  
  rating.rating_type AS rating_type,  
  uniqueid.value AS uniqueid_value,  
  uniqueid.type AS uniqueid_type  
FROM tvshow 
  LEFT JOIN (SELECT idShow, MAX(idPath) as idPath FROM tvshowlinkpath
GROUP BY tvshowlinkpath.idShow) AS tvshowlinkpath ON 
    tvshowlinkpath.idShow=tvshow.idShow 
  LEFT JOIN path ON 
    path.idPath=tvshowlinkpath.idPath 
  INNER JOIN tvshowcounts ON 
    tvshow.idShow = tvshowcounts.idShow  
  LEFT JOIN rating ON 
    rating.rating_id=tvshow.c04  
  LEFT JOIN uniqueid ON 
    uniqueid.uniqueid_id=tvshow.c12 
```

wildcat2020 replied:
Yes, I thought that too, but if we said path information could be eventually ommited, I assumed that correctness of idparentpath should not be an issue...Anyway, however, what you suggest is even better, let me change it that way...

wildcat2020 replied:
Done, please kindly check. I also changed back to select tvshow.* and removed the GROUP BY clause.

wildcat2020 replied:
View with the latest build seems to be working fine...

![image](https://user-images.githubusercontent.com/23155784/50739016-a630ae00-11e2-11e9-9cdc-60db1b41797c.png)

phate89 replied:
You need to bump version to 115 because someone might already have a 114 with yesterday build.. But how did you manage to create a new pr without conflicts starting from the old code? I'm not a git expert..

wildcat2020 replied:
Thanks. I changed the version to 115 now. To tell you the truth, via trial and error... :-) 
I restored my old branch which gave me the old latest code back. Then I changed the file, did git add . , then did git commit, and then I went to the web interface and compared the versions and from there I created a pull request...

wildcat2020 commented:
OK, this now has merge conflicts once I changed to 115 from 114. What versions should I put to resolve this?

fritsch commented:
The hightest.

Am So., 6. Jan. 2019 um 18:08 Uhr schrieb Argyrios Derlopas <
notifications@github.com>:

> OK, this now has merge conflicts once I changed to 115 from 114. What
> versions should I put to resolve this?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/15200#issuecomment-451757590>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABCfHW97DTzuDgU05EohD9W4uAnEDRMDks5vAi1_gaJpZM4Zxobg>
> .
>


-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

wildcat2020 commented:
ok, trying 116...

wildcat2020 commented:
ok, it seems it won't allow anything else than 114...

![image](https://user-images.githubusercontent.com/23155784/50739332-69b38100-11e7-11e9-9a4a-3bd43cebf982.png)

phate89 commented:
I think it's normal.. you have to rebase and resolve all conflicts in your branch..

wildcat2020 commented:
Ok, how do I do that? What I was doing to ensure one commit was the following: 

git checkout feature_myvideos113
git reset master
git add .
git commit 
git push --force origin feature_myvideos113

How do I rebase and apply my change?

Rechi commented:
git remote add upstream https://github.com/xbmc/xbmc.git # only need once if you haven't done that yet

git fetch --all
git checkout feature_myvideos113
git reset upstream/master
 git add .
git commit
git push --force origin feature_myvideos113

wildcat2020 commented:
Many thanks, now it seems it accepts 115 version

Rechi commented on xbmc/video/Teletext.cpp:
```diff
@@ -105,8 +105,8 @@ TextPageAttr_t Text_AtrTable[] =
   { TXT_ColorWhite  , TXT_ColorRed   , C_G0P, 0, 0, 0 ,0, 0, 0, 0, 0, 0, 0, 0x3f}, /* ATR_L250 */
-  { TXT_ColorBlack  , TXT_ColorGreen , C_G0P, 0, 0, 0 ,0, 0, 0, 0, 0, 0, 0, 0x3f}, /* ATR_L251 */
   { TXT_ColorBlack  , TXT_ColorYellow, C_G0P, 0, 0, 0 ,0, 0, 0, 0, 0, 0, 0, 0x3f}, /* ATR_L252 */
+  { TXT_ColorBlack  , TXT_ColorGreen , C_G0P, 0, 0, 0 ,0, 0, 0, 0, 0, 0, 0, 0x3f}, /* ATR_L251 */
```
This change slipped in and has to be removed.

Rechi commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -4899,8 +4901,6 @@ class CShowLink
 void CVideoDatabase::UpdateTables(int iVersion)
 {
-  // Important: DO NOT use CREATE TABLE [...] AS SELECT [...] - it does not work on MySQL with GTID consistency enforced
-
```
same

Rechi commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -5382,15 +5382,9 @@ void CVideoDatabase::UpdateTables(int iVersion)
-    m_pDS->exec("DROP TABLE settingsold");
+    m_pDS->exec("CREATE TABLE settingsnew AS SELECT idFile, Deinterlace, ViewMode, ZoomAmount, PixelRatio, VerticalShift, AudioStream, SubtitleStream, SubtitleDelay, SubtitlesOn, Brightness, Contrast, Gamma, VolumeAmplification, AudioDelay, ResumeTime, Sharpness, NoiseReduction, NonLinStretch, PostProcess, ScalingMethod, DeinterlaceMode, StereoMode, StereoInvert, VideoStream FROM settings");
+    m_pDS->exec("DROP TABLE settings");
+    m_pDS->exec("ALTER TABLE settingsnew RENAME TO settings");
```
same

Rechi commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -7038,7 +7032,6 @@ bool CVideoDatabase::GetMoviesByWhere(const std::string& strBaseDir, const Filte
         std::string path = StringUtils::Format("%i", movie.m_iDbId);
         itemUrl.AppendPath(path);
         pItem->SetPath(itemUrl.ToString());
-        pItem->SetDynPath(movie.m_strFileNameAndPath);
```
same

Rechi commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -7263,28 +7256,27 @@ bool CVideoDatabase::GetEpisodesByWhere(const std::string& strBaseDir, const Fil
       const dbiplus::sql_record* const record = data.at(targetRow);
 
-      CVideoInfoTag episode = GetDetailsForEpisode(record, getDetails);
+      CVideoInfoTag movie = GetDetailsForEpisode(record, getDetails);
```
same

Rechi commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -7263,28 +7256,27 @@ bool CVideoDatabase::GetEpisodesByWhere(const std::string& strBaseDir, const Fil
       if (m_profileManager.GetMasterProfile().getLockMode() == LOCK_MODE_EVERYONE ||
           g_passwordManager.bMasterUser                                     ||
-          g_passwordManager.IsDatabasePathUnlocked(episode.m_strPath, *CMediaSourceSettings::GetInstance().GetSources("video")))
+          g_passwordManager.IsDatabasePathUnlocked(movie.m_strPath, *CMediaSourceSettings::GetInstance().GetSources("video")))
```
same

Rechi commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -7263,28 +7256,27 @@ bool CVideoDatabase::GetEpisodesByWhere(const std::string& strBaseDir, const Fil
+          g_passwordManager.IsDatabasePathUnlocked(movie.m_strPath, *CMediaSourceSettings::GetInstance().GetSources("video")))
       {
-        CFileItemPtr pItem(new CFileItem(episode));
+        CFileItemPtr pItem(new CFileItem(movie));
```
same

Rechi commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -7263,28 +7256,27 @@ bool CVideoDatabase::GetEpisodesByWhere(const std::string& strBaseDir, const Fil
         std::string path;
         if (appendFullShowPath && videoUrl.GetItemType() != "episodes")
-          path = StringUtils::Format("%i/%i/%i", record->at(VIDEODB_DETAILS_EPISODE_TVSHOW_ID).get_asInt(), episode.m_iSeason, idEpisode);
+          path = StringUtils::Format("%i/%i/%i", record->at(VIDEODB_DETAILS_EPISODE_TVSHOW_ID).get_asInt(), movie.m_iSeason, idEpisode);
```
same

Rechi commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -7263,28 +7256,27 @@ bool CVideoDatabase::GetEpisodesByWhere(const std::string& strBaseDir, const Fil
           path = StringUtils::Format("%i", idEpisode);
         itemUrl.AppendPath(path);
         pItem->SetPath(itemUrl.ToString());
-        pItem->SetDynPath(episode.m_strFileNameAndPath);
```
same

Rechi commented on xbmc/video/VideoDatabase.cpp:
```diff
@@ -7263,28 +7256,27 @@ bool CVideoDatabase::GetEpisodesByWhere(const std::string& strBaseDir, const Fil
-        pItem->SetOverlayImage(CGUIListItem::ICON_OVERLAY_UNWATCHED, episode.GetPlayCount() > 0);
-        pItem->m_dateTime = episode.m_firstAired;
+        pItem->SetOverlayImage(CGUIListItem::ICON_OVERLAY_UNWATCHED, movie.GetPlayCount() > 0);
+        pItem->m_dateTime = movie.m_firstAired;
```
same

wildcat2020 replied:
OMG! Are you saying you need me to take out all the green (+) lines and put back all the red (minus) lines?

wildcat2020 replied:
Thanks, done.

wildcat2020 committed:
[database] Correct tvshow_view after PR #15185 merge to make sure it only returns one row per tvshow with correct strPath and idParentPath.

Added subquery in FROM clause in place of tvshowlinkpath table to return only the maximum strPath.

pkerling commented:
@phate89 Cool solution

works for me on SQLite

MartijnKaijser commented:
@phate89 ok for merge?

phate89 commented:
Everything is ok

MilhouseVH commented:
MySQL 5.5.21 server, LibreELEC (latest master):
```
23:29:55.189 T:1817162608  NOTICE: Attempting to update the database MyVideos115 from version 114 to 115
23:30:10.261 T:1817162608   ERROR: SQL: [MyVideos115] Undefined MySQL error: Code (1349)
                                            Query: CREATE VIEW tvshow_view AS SELECT   tvshow.*,  path.idParentPath AS idParentPath,  path.strPath AS strPath,  tvshowcounts.dateAdded AS dateAdded,  lastPlayed, totalCount, watchedcount, totalSeasons,   rating.rating AS rating,   rating.votes AS votes,   rating.rating_type AS rating_type,   uniqueid.value AS uniqueid_value,   uniqueid.type AS uniqueid_type FROM tvshow  LEFT JOIN (SELECT idShow, MAX(idPath) as idPath                FROM tvshowlinkpath               GROUP BY tvshowlinkpath.idShow) AS tvshowlinkpath ON     tvshowlinkpath.idShow=tvshow.idShow  LEFT JOIN path ON    path.idPath=tvshowlinkpath.idPath  INNER JOIN tvshowcounts ON    tvshow.idShow = tvshowcounts.idShow   LEFT JOIN rating ON    rating.rating_id=tvshow.c04   LEFT JOIN uniqueid ON    uniqueid.uniqueid_id=tvshow.c12
23:30:10.262 T:1817162608   ERROR: Exception updating database MyVideos115 from version 114 to 115
```

Pre-MySQL 5.7 doesn't support sub-queries in views: https://stackoverflow.com/questions/23765093/mysql-error-code-1349-views-select-contains-a-subquery-in-the-from-clause?noredirect=1&lq=1

Full migration log: http://ix.io/1y5I

fritsch commented:
Worth a try: https://github.com/fritsch/xbmc/commits/mysql workarounding with DISTINCT ?

pkerling commented:
This is beginning to get on my nerves... You fix it for one MySQL, it breaks on another. And we don't have any infrastructure to even test this. Optimally, that would be part of the test suite. @rechi @kib is it conceivable in any way to have Jenkins access a zoo of MySQL/MariaDB servers (docker containers presumably)? Or is that too complicated? It wouldn't have to be run all the time, just on DB changes (not that frequent).

pkerling commented:
@fritsch I don't think this makes the query predictable

phate89 commented:
We need to create a view with the subquery and then use the view in the query like we do with tvshowcounts...

fritsch commented:
Could you please do that? (It's RC5 time and quit urgent please).

phate89 commented:
I'm on mobile but this afternoon I can do it for sure

pkerling commented:
@phate89 #15235
